subsystems/FPIDSubsystem.java

```diff
@@ -0,0 +1,163 @@
+    protected SpeedController[] motors;
+    protected SpeedControllerGroup motorGroup;
+
+    private final DigitalInput limitSwitch;
+
+    // what type of value we can get from an encoder
+    public static enum EncoderValueType {
+        Velocity, Position
+    }
+
+    public static class FeedforwardModel extends ArmFeedforward {
+        public final double kS;
+        public final double kCos;
+        public final double kV;
+        public final double kA;
+
+        // sets all given values for static, gravity, velocity, and acceleration gains.
+        public FeedforwardModel(double kS, double kCos, double kV, double kA) {
+            super(kS, kCos, kV, kA);
+            this.kS = kS;
+            this.kCos = kCos;
+            this.kV = kV;
+            this.kA = kA;
+        }
+
+        // use this for an elevator, or any FPIDSubsystem that does not require an angle gain.
+        public FeedforwardModel(double kS, double kV, double kA) {
+            this(kS, 0, kV, kA);
+        }
+
+        public double calculateNoAngle(double velocityRadPerSec, double accelRadPerSecSquared) {
+            return (
+                kS * Math.signum(velocityRadPerSec)
+                + kV * velocityRadPerSec
+                + kA * accelRadPerSecSquared
+            );
+        }
+    }
+
+    public static class FPIDConstants {
+        public double speedMultiplier, gearing;
+        public int limitSwitchPort, encoderCPR;
+        public boolean limitSwitchInverted, invertThrottle;
+        public PID pid;
+        public FeedforwardModel ff;
+
+        public FPIDConstants(double speedMultiplier, double gearing, int encoderCPR, boolean limitSwitchInverted, boolean invertThrottle, int limitSwitchPort, PID pid, FeedforwardModel ff) {
+            this.speedMultiplier = speedMultiplier;
+            this.gearing = gearing;
+            this.encoderCPR = encoderCPR;
+            this.limitSwitchPort = limitSwitchPort;
+            this.limitSwitchInverted = limitSwitchInverted;
+            this.invertThrottle = invertThrottle;
+            this.pid = pid;
+            this.ff = ff;
+        }
+    }
+
+    // pass motors as an array of SpeedControllers, eg. WPI_TalonFX[] or CANSparkMax[]
+    public FPIDSubsystem(FPIDConstants constants, SpeedController[] motors) {
+        super(new PIDController(constants.pid.kP, constants.pid.kI, constants.pid.kD));
+        //enable PID control when starting
+        this.enable();
+
+        this.constants = constants;
+
+        limitSwitch = new DigitalInput(constants.limitSwitchPort);
+        motorGroup = new SpeedControllerGroup(motors);
+
+        configureMotors();
+    }
+
+    // have the implementing class configure the motors
+    public abstract void configureMotors();
+
+    // the implementing class also needs to get and set encoder values
+    // param type: what type of value we're getting from the encoder. values Position and Velocity as defined above in EncoderValueType
```

Comment:
change this documentation to be the javadoc standard. here's a guide: https://www.tutorialspoint.com/java/java_documentation.htm